_posts/2020-09-08-Flux3D.md

```diff
@@ -0,0 +1,195 @@
+TriMesh{Float32, UInt32, CUDA.CuArray} Structure:
+    Batch size: 2
+    Max verts: 17443
+    Max faces: 17116
+    offset: -1
+    Storage type: CuArray
+
+julia> save_trimesh("normalize_aiplane.obj",m2,1)
+
+```
+
+### 2. Preparing Dataset
+
+ModelNet10/40 is one of the widely used 3D dataset, and we can easily access and preprocess this dataset for use in various vision-based tasks using Flux3D. We can also use conversion transforms for using ModelNet10/40 in multiple variants like PointCloud, TriMesh, VoxelGrid.
+
+```julia
+julia> using Flux3D, Makie
+
+julia> dset = ModelNet10(categories=["monitor","chair"])
+ModelNet Dataset:
+    root: /root/.julia/packages/Flux3D/9bMbc/datasets/
+    train: true
+    length: 1354
+    transform: nothing
+    categories: 2
+
+julia> dset[1]
+DataPoint:
+    idx: 1
+    data: TriMesh{Float32,UInt32,Array}
+    ground_truth: 1
+    category_name: monitor
+```
+
+We can use conversion transforms to convert the dataset to VoxelGrid. Similarily we can convert VoxelGrid, TriMesh and PointCloud to any 3D structure either using transforms or simply using the regular constructor.
+
+```julia
+julia> t = Chain(NormalizeTriMesh(), TriMeshToVoxelGrid(64))
+Chain(NormalizeTriMesh(;inplace=true), TriMeshToVoxelGrid(resolution=64))
+
+julia> dset2 = ModelNet10(train=false, transform=t)
+dset2 = ModelNet Dataset:
+    root: /root/.julia/packages/Flux3D/9bMbc/datasets/
+    train: false
+    length: 908
+    transform: Chain(NormalizeTriMesh(;inplace=true), TriMeshToVoxelGrid(resolution=64))
+    categories: 10
+
+julia> dset2[1]
+DataPoint:
+    idx: 1
+    data: VoxelGrid{Float32}
+    ground_truth: 1
+    category_name: bathtub
+```
+
+### 3. Defining Model/Loss
+
+Loss objectives and designing model solely depend upon the requirement of tasks, and with the help of FLuxML ecosystem, we can define any custom model as well as loss. There are some commonly used metrics and predefined 3D models which can assist in 3D specific tasks like chamfer_distance, laplacian_loss and edge_loss. 
+
+```julia
+julia> chamfer_distance(p,p)
+1.58147f-5
+
+julia> laplacian_loss(m)
+6.72288
+
+julia> edge_loss(m)
+2110.6
+```
+
+### 4. Training
+
+Flux and Zygote takes care of everything here :)
+
+Additonally, 3D structures and all relevant transforms, as well as metrics, are compatible with Zygote for supporting differentiation.
+
+### 5. Visualization and Evaluation
+
+Flux3D provides a function visualize for visualizing 3D structures. This function uses Makie for plotting. We can use this same function for visualizing all three 3D structures PointCloud, TriMesh, and VoxelGrid
+
+```julia
+
+julia> using Makie
+
+julia> vbox(
+    hbox(
+        visualize(m,1),visualize(p,1,markersize=25)
+        ),
+    hbox(
+        visualize(v,1), visualize(v,1,algo=:MarchingCubes)
+        ),
+    )
+
+```
+
+<div style="text-align:center">
+  <img src="/assets/2020-09-08-Flux3D/visualize.png">
+</div>
+
+## Benchmarks for Flux3D.jl
```

Comment:
I think the benchmarks need to be higher up, with a short sentence talking about how it compares with kaolin.

Reply:
Next paragraph briefly talks about the comparison with the kaolin.

Reply:
The message that this approach makes sense is somewhat lost with the real results from the benchmarks being very low down. Maybe just have a sentence here and say that it is discussed further in the post.

Reply:
made necessary changes